os/binfmt/libelf/libelf_bind.c

```diff
@@ -277,6 +281,69 @@ static int elf_relocateadd(FAR struct elf_loadinfo_s *loadinfo, int relidx, FAR
 	return -ENOSYS;
 }
 
+#ifdef CONFIG_SUPPORT_COMMON_BINARY
+static int export_library_symtab(FAR struct elf_loadinfo_s *loadinfo)
+{
+	int i;
+	int ret;
+
+	Elf32_Shdr *symtab = &loadinfo->shdr[loadinfo->symtabidx];
+	int nsyms = symtab->sh_size / sizeof(Elf32_Sym);
+
+	g_lib_symtab = kmm_malloc(nsyms * sizeof(struct symtab_s));
+	if (!g_lib_symtab) {
+		berr("Failed to allocate memory for exporting symbol table\n");
+		return -ENOMEM;
+	}
+
+	g_num_lib_syms = 0;
+	for (i = 0; i < nsyms; i++) {
+		Elf32_Sym sym;
+		ret = elf_readsym(loadinfo, i, &sym);
+		if (ret < 0) {
+			berr("Failed to read symbol[%d]: %d\n", i, ret);
+			goto ret_err;
+		}
+
+		if (ELF32_ST_BIND(sym.st_info) == STB_GLOBAL) {
+			ret = elf_symname(loadinfo, &sym);
+			if (ret < 0) {
+				berr("SHN_UNDEF: Failed to get symbol name: %d\n", ret);
+				goto ret_err;
+			}
+
+			g_lib_symtab[g_num_lib_syms].sym_name = kmm_zalloc(strlen(loadinfo->iobuffer) + 1);
+			if (!g_lib_symtab[g_num_lib_syms].sym_name) {
+				berr("Error allocating symbol name %d entry\n", g_num_lib_syms);
+				ret = -ENOMEM;
+				goto ret_err;
+			}
+
+			strcpy(g_lib_symtab[g_num_lib_syms].sym_name, loadinfo->iobuffer);
+			ret = elf_symvalue(loadinfo, &sym, 0, 0);
+			if (ret < 0) {
+				if (ret == -ESRCH) {
+					berr("Undefined symbol[%d] has no name: %d\n", i, ret);
+				} else {
+					berr("Failed to get value of symbol[%d]: %d\n", i, ret);
+				}
+				goto ret_err;
+			}
+			g_lib_symtab[g_num_lib_syms].sym_value = sym.st_value;
+			g_num_lib_syms++;
+		}
+	}
+
+	binfo("Total symbols in library = %d. Exported symbols = %d\n", nsyms, g_num_lib_syms);
+	kmm_realloc(g_lib_symtab, g_num_lib_syms * sizeof(struct symtab_s));
+	return OK;
+
+ret_err:
+	kmm_free(g_lib_symtab);
```

Comment:
Is it okay not to free the `g_lib_symtab[g_num_lib_syms].sym_name`?

Reply:
It needs to be freed. Fixed.